back/Dockerfile

```diff
@@ -1,31 +1,57 @@
+# gemspec files) from the project. This enables us to optimize Docker layer caching by
+# installing all the dependencies before copying the remaining project files to the
+# Docker image in the later stages.
+FROM alpine:3.17.2 AS dependency_files
+SHELL ["/bin/ash", "-o", "pipefail", "-c"]
 
-ARG CITIZENLAB_EE=false
-ENV CITIZENLAB_EE=$CITIZENLAB_EE
+WORKDIR /app
 
-RUN apt-get update && apt-get install -qq -y --no-install-recommends \
-      build-essential libpq-dev file imagemagick curl git optipng jpegoptim pngquant libgeos-dev libgmp3-dev netcat shared-mime-info \
-      less
+COPY back/Gemfile back/Gemfile.* ./
+COPY back/engines ./engines
 
-RUN curl -sL https://deb.nodesource.com/setup_15.x  | bash -
-RUN apt-get -y install nodejs
+# `version.rb` files are also kept as they are required by gemspec files.
+RUN find engines -type f -not -name "*.gemspec" -not -name "version.rb" -print0 | xargs -0 rm -rf \
+ && find engines -type d -empty -delete
 
-# Install MJML parser required by email engine.
-RUN npm install -g mjml@4.4.1
+FROM ruby:2.7.6-slim
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
-ENV INSTALL_PATH /cl2_back
-RUN mkdir -p $INSTALL_PATH
-WORKDIR $INSTALL_PATH
+ARG CITIZENLAB_EE=false
+ENV CITIZENLAB_EE=$CITIZENLAB_EE
 
-COPY back/Gemfile back/Gemfile.* ./
-COPY back/engines ./engines
-COPY back/lib/citizen_lab.rb ./lib/
+RUN apt-get update                                               \
+ && apt-get install -qq -y --no-install-recommends               \
+      build-essential                                            \
+      curl                                                       \
+      file                                                       \
+      git                                                        \
+      imagemagick                                                \
+      jpegoptim                                                  \
+      less                                                       \
+      libgeos-dev                                                \
+      libgmp3-dev                                                \
+      libpq-dev                                                  \
+      netcat                                                     \
+      optipng                                                    \
+      pngquant                                                   \
+      shared-mime-info                                           \
+ && curl -sL https://deb.nodesource.com/setup_18.x  | bash -     \
+ && apt-get -y --no-install-recommends install nodejs            \
+    # Install MJML parser required by email engine.
+ && npm install -g mjml@4.4.1                                    \
+ && apt-get clean                                                \
+ && rm -rf /var/lib/apt/lists/*
+
+WORKDIR /cl2_back
 
 RUN gem install bundler:2.3.20
+
+COPY --from=dependency_files /app .
 RUN bundle install
```

Comment:
Maybe reverting this commit could make it faster https://github.com/CitizenLabDotCo/citizenlab/commit/778d95f
Because I believe `bundle install` is not CPU-bound, but network-bound.
On default medium, the number of CPU is 2. Probably, using 4 jobs can speed it up.

But maybe should be tested on CI first.

Reply:
Good point! I'll try that out in the second part.